packages/core/src/render3/i18n/i18n_parse.ts

```diff
@@ -62,148 +52,186 @@ function replaceNgsp(value: string): string {
-                icuNodeIndex << I18nUpdateOpCode.SHIFT_REF | I18nUpdateOpCode.IcuSwitch, tIcuIndex,
-                mask,  // mask of all the bindings of this ICU expression
-                2,     // skip 2 opCodes if not changed
-                icuNodeIndex << I18nUpdateOpCode.SHIFT_REF | I18nUpdateOpCode.IcuUpdate, tIcuIndex);
-          } else if (parts[j] !== '') {
-            const text = parts[j] as string;
-            // Even indexes are text (including bindings)
-            const hasBinding = text.match(BINDING_REGEXP);
-            // Create text nodes
-            const textNodeIndex = allocNodeIndex(startIndex);
-            createOpCodes.push(
-                // If there is a binding, the value will be set during update
-                hasBinding ? '' : text, textNodeIndex,
-                parentIndex << I18nMutateOpCode.SHIFT_PARENT | I18nMutateOpCode.AppendChild);
-
-            if (hasBinding) {
-              addAllToArray(generateBindingUpdateOpCodes(text, textNodeIndex), updateOpCodes);
-            }
-          }
-        }
+        const tNode = createTNodePlaceholder(tView, existingTNodeStack[0], index);
+        existingTNodeStack.unshift([]);
+        setCurrentTNode(tNode, true);
       }
     }
   }
 
-  if (i18nVarsCount > 0) {
-    allocExpando(tView, lView, i18nVarsCount);
-  }
-
-  // NOTE: local var needed to properly assert the type of `TI18n`.
-  const tI18n: TI18n = {
-    vars: i18nVarsCount,
+  tView.data[index + HEADER_OFFSET] = <TI18n>{
     create: createOpCodes,
     update: updateOpCodes,
-    icus: icuExpressions.length ? icuExpressions : null,
   };
+}
 
-  tView.data[index + HEADER_OFFSET] = tI18n;
+/**
+ * Allocate space in i18n Range add create OpCode instruction to crete a text or comment node.
+ *
+ * @param tView Current `TView` needed to allocate space in i18n range.
+ * @param rootTNode Root `TNode` of the i18n block. This node determines if the new TNode will be
+ *     added as part of the `i18nStart` instruction or as part of the `TNode.insertBeforeIndex`.
+ * @param existingTNodes internal state for `addTNodeAndUpdateInsertBeforeIndex`.
+ * @param lView Current `LView` needed to allocate space in i18n range.
+ * @param createOpCodes Array storing `I18nCreateOpCodes` where new opCodes will be added.
+ * @param text Text to be added when the `Text` or `Comment` node will be created.
+ * @param isICU true if a `Comment` node for ICU (instead of `Text`) node should be created.
+ */
+function createTNodeAndAddOpCode(
+    tView: TView, rootTNode: TNode|null, existingTNodes: TNode[], lView: LView,
+    createOpCodes: I18nCreateOpCodes, text: string, isICU: boolean): TNode {
+  const i18nNodeIdx = allocExpando(tView, lView, 1);
+  let opCode = i18nNodeIdx << I18nCreateOpCode.SHIFT;
+  const previousOrParent = getCurrentTNode()!;
+  const isParent = isCurrentTNodeParent();
+  let parentTNode = isParent ? previousOrParent : previousOrParent.parent;
```

Comment:
This can probably be replaced by `getCurrentParentTNode` call:
```suggestion
  let parentTNode = getCurrentParentTNode();
```